diff --git a/secded_cfg.hjson b/secded_cfg.hjson
index d1aba6db1..5de93ce7d 100644
--- a/secded_cfg.hjson
+++ b/secded_cfg.hjson
@@ -2,6 +2,7 @@
 // Licensed under the Apache License, Version 2.0, see LICENSE for details.
 // SPDX-License-Identifier: Apache-2.0
 { cfgs: [
+    {k: 8,    m: 5,  code_type: "hsiao"  },
     {k: 16,   m: 6,  code_type: "hsiao"  },
     {k: 22,   m: 6,  code_type: "hsiao"  },
     {k: 32,   m: 7,  code_type: "hsiao"  },
